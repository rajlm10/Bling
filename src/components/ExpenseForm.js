import React from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import {SingleDatePicker} from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'


export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props)
        this.state={
            description:props.expense?props.expense.description : '',
            note:props.expense?props.expense.note:'',
            amount:props.expense?props.expense.amount.toString():'',
            createdAt:props.expense?moment(props.expense.createdAt):moment(),
            calenderFocused:false,
            error:''
        }
    }
   
    onDescriptionChange=(e)=>{
        const description=e.target.value
        this.setState(()=>{
            return{
                description
            }
        })
    }
    onNoteChange=(e)=>{
        const note=e.target.value
        this.setState(()=>{
            return{
                note
            }
        })
    }
    onAmountChange=(e)=>{
        const amount=e.target.value
        if (!amount|| amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>{
                return{
                    amount
                }
            })
        }

    }
    onDateChange=(createdAt)=>{
        if (createdAt) {
            this.setState(()=>{
                return{
                    createdAt
                }
            })   
        }

    }
    onFocusChange=({focused})=>{
        this.setState(()=>{
            return{
                calenderFocused:focused
            }
        })
    }
    onSubmit=(e)=>{
        e.preventDefault()
        if (!this.state.description||!this.state.amount) {
            this.setState(()=>{
                return{
                    error:'Please provide a description and amount'
                }
            })
        }
        else{
            this.setState(()=>{
                return{
                    error:''
                }
            })
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount),
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note

            })
        }
    }
    render(){
        return(
            <div>
            {this.state.error&&<p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>
                <input type="text" placeholder="Description" autoFocus
                value={this.state.description} onChange={this.onDescriptionChange}
                />

                <input type="number" placeholder="Amount" 
                value={this.state.amount} onChange={this.onAmountChange}
                />

                <SingleDatePicker 
                date={this.state.createdAt}
                onDateChange={this.onDateChange}
                focused={this.state.calenderFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={()=>false}
                
                />

                <textarea placeholder="Any notes you'd like to add?"
                value={this.state.note} onChange={this.onNoteChange}
                >
                </textarea>

                <button>Add Expense</button>
            </form>
            </div>
        )
    }
}