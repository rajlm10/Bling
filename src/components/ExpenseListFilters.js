import React from 'react'
import {connect} from 'react-redux'
import {setTextFilter,sortByAmount,sortByDate,setStartDate,setEndDate} from '../actions/filters'
import {DateRangePicker} from 'react-dates'


class ExpenseListFilters extends React.Component{
    state={
        calendarFocused:null
    }
    onDatesChange=({startDate,endDate})=>{
        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    }
    onFocusChange=(calendarFocused)=>{
        this.setState(()=>{
            return{
                calendarFocused
            }
        })
    }
    render(){
        return(
            <div>
                <input type="text" value={this.props.filters.text} onChange={
                    (e)=>{
                        this.props.dispatch(setTextFilter(e.target.value))
                    }
                }/>
                <select value={this.props.filters.sortBy} onChange={
                    (e)=>{
                        if (e.target.value==='date') {
                            this.props.dispatch(sortByDate())
                        }
    
                        else if(e.target.value==='amount'){
                            this.props.dispatch(sortByAmount())
                        }
                        
                    }
                }>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                startDate={this.props.filters.startDate}
                startDateId="start"
                endDateId="end"
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={()=>false}
                showClearDates={true}
                
                />
    
            </div>
        )
    }
}



const mapStateToProps=(state)=>{
    return{
        filters:state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListFilters)