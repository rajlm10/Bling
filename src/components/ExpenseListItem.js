import React from 'react'
import { Link } from 'react-router-dom'

const ExpenseListItem=(props)=>{
    return(
        <div>
            <h3><Link to={`/edit/${props.expense.id}`}>{props.expense.description}</Link></h3>
            <p>{props.expense.amount} - {props.expense.createdAt}</p>

        </div>
    )
}

export default (ExpenseListItem)