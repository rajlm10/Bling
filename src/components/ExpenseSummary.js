import React from 'react'
import {Link} from 'react-router-dom'
import { connect } from 'react-redux'
import selectExpenses from '../selectors/expenses'
import selectTotal from '../selectors/expensesTotal'

const ExpenseSummary=(props)=>{
    return(
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{props.expenses.length}</span> expenses,totalling <span>{props.total}</span></h1>
                <div className="page-header__actions">
                    <Link to="/create" className="button">Add Expense</Link>
                </div>
            </div>
        </div>
    )
}

const mapStateToProps=(state)=>{
    return{
        expenses:selectExpenses(state.expenses,state.filters),
        total:selectTotal(selectExpenses(state.expenses,state.filters))
    }
}

export default connect(mapStateToProps)(ExpenseSummary)