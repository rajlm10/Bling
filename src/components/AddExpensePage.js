import React from 'react'
import {connect} from 'react-redux'
import {startAddExpense} from '../actions/expenses'
import ExpenseForm from './ExpenseForm'
const AddExpensePage=(props)=>{
    
    
    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <div className="page-header__title">
                        <h1>What did you spend it on?</h1>
                    </div>
                </div>
            </div>

            <div className="content-container">
                <ExpenseForm 
                onSubmit={(expense)=>{
                    props.dispatch(startAddExpense(expense))
                    props.history.push('/')
                    
                }}
                />
            </div>

        </div>
    )
}

export default connect()(AddExpensePage)