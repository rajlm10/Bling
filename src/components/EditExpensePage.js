import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {startEditExpense,startRemoveExpense} from '../actions/expenses'

const EditExpensePage=(props)=>{
    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Like to make changes?</h1>
                </div>
            </div>
            <div className="content-container">
                <ExpenseForm
                expense={props.expense}
                onSubmit={(expense)=>{
                    props.dispatch(startEditExpense(props.expense.id,expense))
                    props.history.push('/')
                    
                }}
                
                />
                <button onClick={()=>{
                    props.dispatch(startRemoveExpense({id:props.expense.id}))
                    props.history.push('/')
                }}
                className="button button__remove"
                >
                Remove
                </button>
            </div>
        </div>
    )
}

const mapStateToProps=(state,props)=>{
    return{
        expense:state.expenses.find((expense)=>{
            return expense.id===props.match.params.id
        })
    }
}

export default connect(mapStateToProps)(EditExpensePage)