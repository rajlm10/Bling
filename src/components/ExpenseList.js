import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'


export const ExpenseList=(props)=>{
    return(
        <div>
            {
                props.expenses.length===0&&<p>Please add an expense</p>
            }
                
            {props.expenses.map((expense)=>{
                return(
                    
                    <ExpenseListItem key={expense.id} expense={expense} />
                    
                )
            })}
        </div>
    )
}

const mapStatetoProps=(state)=>{
    
    
    return{
        expenses:selectExpenses(state.expenses,state.filters)
    }
}

export default connect(mapStatetoProps)(ExpenseList)


//Here react redux helps us by rerendering ExpenseList every time the store changes, connect joins the component to the state of the
//store