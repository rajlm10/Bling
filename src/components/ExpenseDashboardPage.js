import React from 'react'
import ExpenseSummary from './ExpenseSummary'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
const ExpenseDashboardPage=()=>{
    return (
        <div>
            <ExpenseSummary/>
            <ExpenseListFilters/>
            <ExpenseList />
        </div>
    )
}

export default ExpenseDashboardPage