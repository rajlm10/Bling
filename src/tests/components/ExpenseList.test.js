import React from 'react'
import {shallow} from 'enzyme'
import {ExpenseList} from '../../components/ExpenseList'
import moment from 'moment'

const expenses=[{
    id:1,
    description:'Gum',
    note:'',
    amount:195,
    createdAt:0
},
{
    id:2,
    description:'Rent',
    note:'',
    amount:109500,
    createdAt:moment(0).subtract(4,'days').valueOf()
},
{
    id:3,
    description:'Credit Card',
    note:'',
    amount:4500,
    createdAt:moment(0).add(4,'days').valueOf()
}]

test('should render expense list with expenses',()=>{
    const wrapper=shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should render message  with no expenses',()=>{
    const wrapper=shallow(<ExpenseList expenses={[]}/>)
    expect(wrapper).toMatchSnapshot()
})