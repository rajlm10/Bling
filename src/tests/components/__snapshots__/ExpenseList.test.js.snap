// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list with expenses 1`] = `
<div>
  <ExpenseListItem
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": 1,
        "note": "",
      }
    }
    key="1"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": 2,
        "note": "",
      }
    }
    key="2"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": 3,
        "note": "",
      }
    }
    key="3"
  />
</div>
`;

exports[`should render message  with no expenses 1`] = `
<div>
  <p>
    Please add an expense
  </p>
</div>
`;
